In [8]:
import pandas as pd
from pprint import pprint
import numpy as np
from IPython.display import clear_output

In [9]:
pd.set_option('display.max_columns', 500)
pd.set_option('display.max_colwidth', 100)

In [10]:
df = pd.read_pickle("nyt_midway")

In [11]:
df.shape

(1226, 26)

In [12]:
df.sample()

_id abstract blog  \
1430  5454145f38f0d839202e9940      NaN   {}   

                                                                                                   byline  \
1430  {'original': 'By JEREMY W. PETERS', 'person': [{'firstname': 'Jeremy', 'middlename': 'W.', 'last...   

     document_type  \
1430       article   

                                                                                                 headline  \
1430  {'main': 'In Albany, No Action on School Control for Bloomberg', 'kicker': None, 'content_kicker...   

                                                                                                 keywords  \
1430  [school, fails, senate, bill, democrats, law, mayor, mayors, bloombergs, york, walcott, control,...   

                                                                                               multimedia  \
1430  [{'rank': 0, 'subtype': 'xlarge', 'caption': None, 'credit': None, 'type': 'image', 'url': 'imag...   

     new_desk news_desk print_page              pub_date         score  \
1430    Metro       NaN         18  2009-07-17T00:00:00Z  2.814669e-12   

     section_name slideshow_credits  \
1430          NaN               NaN   

                                                                                                  snippet  \
1430  Albany Democrats blamed New York City&#8217;s mayor for refusing to accept changes, while his ai...   

                  source type_of_material  uri  \
1430  The New York Times             News  NaN   

                                                        web_url  word_count  \
1430  https://www.nytimes.com/2009/07/17/nyregion/17albany.html         800   

                  clean_pubdate  year  month  \
1430  2009-07-17T00:00:00+00:00  2009      7   

                                                                                                full_text  \
1430  Mr. Bloomberg’s aides and Republican allies accused Democrats of waiting until the last minute t...   

                                                                                                  summary  
1430  “One hundred and fifty thousand employees, 1,500 schools, 1.1 million children being held up — f...

In [13]:
df_clean = df.drop_duplicates(subset=['snippet','web_url','clean_pubdate'])

In [14]:
df_clean.shape

(864, 26)

Need to extract useful information from following columns (which are dictionaries)
* Byline - 'original' subtract 'by'
* Headline - 'main'

In [15]:
for idx, row in df_clean.iterrows():
    try:
        byline = row['byline']
        print(byline)
        print(f"Type Byline: {type(byline)}")

        author = row['byline']['original']
        print(f"author: {author}")
        df_clean.loc[idx,'author']=author.replace('By ','')
    except:
        df_clean.loc[idx,'author']='no_author'

clear_output()
print('clear output')

clear output


In [16]:
df_clean.sample()

_id  \
11025  5407c3e238f0d843936a7172   

                                                                                                  abstract  \
11025  Christine C Quinn, former speaker of New York City Council, joins board of abortion-rights group...   

      blog  \
11025   {}   

                                                                                                    byline  \
11025  {'original': 'By THOMAS KAPLAN', 'person': [{'firstname': 'Thomas', 'middlename': None, 'lastnam...   

      document_type  \
11025       article   

                                                                                                  headline  \
11025  {'main': 'Back in Politics, Quinn Will Be Adviser on Abortion-Rights Strategy', 'kicker': None, ...   

                                                                                                  keywords  \
11025  [work, really, politics, ms, dont, abortionrights, primary, place, proud, quinn, quinns, strateg...   

                                                                                                multimedia  \
11025  [{'rank': 0, 'subtype': 'wide', 'caption': None, 'credit': None, 'type': 'image', 'url': 'images...   

      new_desk news_desk print_page              pub_date         score  \
11025    Metro       NaN         23  2014-09-04T00:00:00Z  2.151073e-08   

      section_name slideshow_credits  \
11025          NaN               NaN   

                                                                                                   snippet  \
11025  Christine C. Quinn, the former City Council president, has joined the board of Naral Pro-Choice ...   

                   source type_of_material  uri  \
11025  The New York Times             News  NaN   

                                                                                                   web_url  \
11025  https://www.nytimes.com/2014/09/04/nyregion/back-in-politics-quinn-will-be-adviser-on-abortion-r...   

       word_count              clean_pubdate  year  month  \
11025         687  2014-09-04T00:00:00+00:00  2014      9   

                                                                                                 full_text  \
11025  While Ms. Quinn, 48, has spent most of this year out of the public view, in her free time she ha...   

                                                                                                   summary  \
11025  While Ms. Quinn, 48, has spent most of this year out of the public view, in her free time she ha...   

              author  
11025  THOMAS KAPLAN

In [17]:
def get_headline(row):
    try:
        headline = row['headline']['main']
        print(f"Headline: {headline}")
        print(f"Type: {type(headline)}")
        row['headline_clean'] = headline
        return row
    except:
        row['headline_clean'] = 'no_headline'
        return row

df_clean = df_clean.apply(get_headline,axis=1)

Headline: Bill Banning Gov. Cuomo From Taking Appointee Donations Clears Senate
Type: <class 'str'>
Headline: Bill Banning Gov. Cuomo From Taking Appointee Donations Clears Senate
Type: <class 'str'>
Headline: NY Senate OKs Bill That Would Fix Spelling in Bridge's Name
Type: <class 'str'>
Headline: Details of Counts Against Ex-Senate Leader
Type: <class 'str'>
Headline: In Closing Arguments, Bruno as Bully or Honest Worker
Type: <class 'str'>
Headline: Whistling Past the Deficit
Type: <class 'str'>
Headline: For Bruno, Politics and Business Intersected on the Golf Course
Type: <class 'str'>
Headline: Jump-Start the New York Senate
Type: <class 'str'>
Headline: Blame Panic in G.O.P. for Standoff in Albany
Type: <class 'str'>
Headline: Monserrate’s Flip Creates Tie in New York Senate
Type: <class 'str'>
Headline: Justice on the Farm
Type: <class 'str'>
Headline: One Senate, Two Gavels
Type: <class 'str'>
Headline: Pedro Espada Jr. In Action
Type: <class 'str'>
Headline: Get Back to Work,

Headline: Rooting Out Corruption in Albany
Type: <class 'str'>
Headline: The New York Convention Con
Type: <class 'str'>
Headline: Cuomo as Leader: Focused, Intense and at Times Alienating
Type: <class 'str'>
Headline: Injustice for Children
Type: <class 'str'>
Headline: A Carl Paladino Volunteer Shares the Anger
Type: <class 'str'>
Headline: A Cuomo Heads Back to the Executive Mansion
Type: <class 'str'>
Headline: What Is Albany Doing? Silver Says, &#8216;Quite a Lot&#8217;
Type: <class 'str'>
Headline: Cuomo Accepts Millions From Interests He Assails
Type: <class 'str'>
Headline: The Wrong Way and the Right Way
Type: <class 'str'>
Headline: Taxes for Your Health. Yeah, Sure.
Type: <class 'str'>
Headline: Trial Lawyer Aims to Be Top State Prosecutor
Type: <class 'str'>
Headline: Trial Lawyer Aims to Be Top State Prosecutor
Type: <class 'str'>
Headline: What Are They Doing?
Type: <class 'str'>
Headline: Blame Albany
Type: <class 'str'>
Headline: On the Road with Ed Koch
Type: <class 's

Headline: 2 Lawsuits Detail Abuse of Disabled in State Care
Type: <class 'str'>
Headline: Want to Read Cuomo&#8217;s Story? Biographers Will Offer Options
Type: <class 'str'>
Headline: Angry at Pension Vote, Unions Plan to Skip Hispanic Event
Type: <class 'str'>
Headline: Songs Against Drilling
Type: <class 'str'>
Headline: A Club Refutes the Myth Black People Don’t Ski
Type: <class 'str'>
Headline: Mr. Cuomo&#8217;s New Year
Type: <class 'str'>
Headline: The Nubian Empire Ski Club
Type: <class 'str'>
Headline: The Week in Pictures for Feb. 10
Type: <class 'str'>
Headline: New York, Take Note
Type: <class 'str'>
Headline: Cuomo’s Office Is Said to Rein In Ethics Board He Created
Type: <class 'str'>
Headline: Nonprofits Are Balking at Law on Disclosing Political Donors
Type: <class 'str'>
Headline: The Many Faces of State Political Scandals
Type: <class 'str'>
Headline: Capitol Redux
Type: <class 'str'>
Headline: As Holy Days Dot Calendar, State Budget Must Wait
Type: <class 'str'>
Head

Headline: Thomas Libous, Powerful New York Senator Felled by Scandal, Dies at 63
Type: <class 'str'>
Headline: Legislators Seek to Promote Diversity at Elite Public High Schools
Type: <class 'str'>
Headline: An Austerity Plan That Would Punish the City
Type: <class 'str'>
Headline: Transcript of Cuomo’s 2016 State of the State Address
Type: <class 'str'>
Headline: Racism Charges in Bus Incident, and Their Unraveling, Upset University at Albany
Type: <class 'str'>
Headline: Surveillance Video on City Bus
Type: <class 'str'>
Headline: NY Regulators Order Full Environmental Review for Pipeline
Type: <class 'str'>
Headline: A Corruption Case Worth Repeating
Type: <class 'str'>
Headline: No Deal, Just Blame, on Mayoral Control of New York City Schools
Type: <class 'str'>
Headline: A Streamlined Way to Build Projects Runs Into New York Politics
Type: <class 'str'>
Headline: Andrew Cuomo to Call Special Session to Extend Mayoral Control of Schools
Type: <class 'str'>
Headline: Tired of the Su

Headline: New York Rent Regulation Laws Expire Amid Last-Minute Talks
Type: <class 'str'>
Headline: Time for New York’s Legislature to Address Important Issues
Type: <class 'str'>
Headline: New York State Legislature Will Convene Amid a Changed Political Landscape
Type: <class 'str'>
Headline: Pushing Ethics Reforms, Cuomo Wants Lawmakers to Reveal Income or Lose It
Type: <class 'str'>
Headline: M.T.A. Official Warns Board That Fare and Toll Increases May Be Needed
Type: <class 'str'>
Headline: Poll Finds Little Support for Cuomo’s Policy on Purging State Emails
Type: <class 'str'>
Headline: Gov. Cuomo Talks Reform
Type: <class 'str'>
Headline: Demands to Rename Donald J. Trump State Park Gain Ground
Type: <class 'str'>
Headline: Marching Toward a $15 Minimum Wage
Type: <class 'str'>
Headline: Cuomo’s Proposal for an Executive-Branch Prosecutor Is Called a ‘Scheme’
Type: <class 'str'>
Headline: Official Defends Cuomo Program That Has Created Only 408 Jobs
Type: <class 'str'>
Headline: 

Headline: Failing Grade
Type: <class 'str'>
Headline: New York’s Antique Divorce Law
Type: <class 'str'>
Headline: Experiencing Las Vegas, in Queens
Type: <class 'str'>
Headline: Nice Work if You Can Get It
Type: <class 'str'>
Headline: Groups Spent $1.8 Million in Same-Sex Marriage Push
Type: <class 'str'>
Headline: They Need to Stand Up for Equality
Type: <class 'str'>
Headline: After Talks With G.O.P., Cuomo Expects Passage of Gay Marriage Bill
Type: <class 'str'>
Headline: With No Decision on Same-Sex Marriage, Lawmakers Adjourn for Night
Type: <class 'str'>
Headline: Mom-and-Pop Pharmacy Bill
Type: <class 'str'>
Headline: Old Allies Reunite In a Debate On Layoffs
Type: <class 'str'>
Headline: Governor Cuomo’s List
Type: <class 'str'>
Headline: New York State Weighs Tax Change for BlackRock
Type: <class 'str'>
Headline: Within Our Means
Type: <class 'str'>
Headline: New York’s Thursday Primary
Type: <class 'str'>
Headline: In Karaoke, Lawmakers Find a Catalyst for Collegiality
Type

In [18]:
df_clean.sample()

_id abstract blog byline document_type  \
12628  53c6fabc38f0d858f3e3839a      NaN   {}    NaN       article   

                                                                                                  headline  \
12628  {'main': 'Paid Notice: Deaths  GOODMAN, ROY M', 'kicker': None, 'content_kicker': None, 'print_h...   

                                                                                                  keywords  \
12628  [education, york, president, goodman, college, trustees, brgoodman, m, state, goodmans, deaths, ...   

      multimedia    new_desk news_desk print_page              pub_date  \
12628         []  Classified       NaN         08  2014-06-09T00:00:00Z   

              score section_name slideshow_credits  \
12628  5.841962e-08          NaN               NaN   

                                                                                                   snippet  \
12628  GOODMAN--Roy M. The President and Trustees of Teachers College mourn former State Senator Roy M....   

                   source   type_of_material  uri  \
12628  The New York Times  Paid Death Notice  NaN   

                                                                                  web_url  \
12628  https://query.nytimes.com/gst/fullpage.html?res=9407E1D71E3AF93AA35755C0A9629D8B63   

       word_count              clean_pubdate  year  month  \
12628         124  2014-06-09T00:00:00+00:00  2014      6   

                                                                                                 full_text  \
12628  GOODMAN--Roy M. The President and Trustees of Teachers College mourn former State Senator Roy M....   

                                                                                                   summary  \
12628  GOODMAN--Roy M. The President and Trustees of Teachers College mourn former State Senator Roy M....   

          author                       headline_clean  
12628  no_author  Paid Notice: Deaths  GOODMAN, ROY M

In [19]:
df_clean[df_clean['headline_clean']=='no_headline']

Empty DataFrame
Columns: [_id, abstract, blog, byline, document_type, headline, keywords, multimedia, new_desk, news_desk, print_page, pub_date, score, section_name, slideshow_credits, snippet, source, type_of_material, uri, web_url, word_count, clean_pubdate, year, month, full_text, summary, author, headline_clean]
Index: []

In [20]:
def editorial_author(row):
    if row['author']=='no_author' and row['new_desk']=='Editorial':
        row['author'] = 'THE EDITORIAL BOARD'
    return row

In [21]:
df_clean = df_clean.apply(editorial_author,axis=1)

In [22]:
df_clean.loc[(df_clean['author']=='THE EDITORIAL BOARD')]

_id  \
1045   5458808338f0d8086c45e650   
1091   5452dd0938f0d81b603ed2b7   
1118   5455f05c38f0d84a08fb2ca6   
1219   5453340938f0d81b603eda60   
1396   5469414d38f0d8317bec0cec   
3057   5421dbb238f0d83c6e5b7569   
3575   5442b4d038f0d8246e88c395   
3598   5438033138f0d83c143b7e9a   
7164   515da3e2cf28d0219d001622   
45     5af3ae9f068401528a2ad3f7   
466    5a65463810f40f00018bd220   
992    5459016738f0d8086c45f16b   
995    5458ceb938f0d8086c45ecf8   
1028   545a4ce038f0d82ce42689ed   
1076   54576bd338f0d86ef22a1a3a   
1174   5454d9b538f0d839202eaa19   
1187   5453debb38f0d839202e942b   
1236   5452bb2338f0d81b603ecf91   
1352   546a371838f0d84ca2c5beb9   
1403   545bac7a38f0d85e5ee9100c   
1428   546897c038f0d81e5cd0dbf0   
1497   546843b838f0d81e5cd0d4bf   
1528   5458c47d38f0d8086c45ec1e   
1573   54583d5a38f0d8086c45e0c6   
1740   54566a8a38f0d86ef22a03e6   
1756   54563ead38f0d84a08fb335f   
1844   5472304538f0d85d607d5254   
1866   5472123438f0d847c5e43b5a   
2203   546aef5c38f0d84ca2c5cf8a   
2282   546a949338f0d84ca2c5c76e   
...                         ...   
1164   54538b4e38f0d81b603ee228   
1239   54539ea938f0d81b603ee3d9   
3836   5445570d38f0d832c813ea1e   
5084   540f9e0338f0d87641c67f7d   
11231  5383e0a338f0d84ab09bb06c   
1072   5457c48238f0d8086c45d630   
1253   5452a6d038f0d81b603ecd83   
1284   546a677638f0d84ca2c5c356   
2127   5452e78838f0d81b603ed3a5   
3082   54463f1e38f0d832c813fdfe   
3201   5417e5d538f0d851fd3a89ea   
3446   544454ca38f0d82782d9b73a   
4127   5417da1a38f0d851fd3a88d9   
4324   54461b0338f0d832c813fada   
4614   5443d8c538f0d8246e88dba7   
4784   5443216a38f0d8246e88cc5e   
4908   54380b9b38f0d83c143b813d   
5047   540ec71138f0d863ce5a5383   
5115   540eb7cb38f0d863ce5a5258   
5520   540156f538f0d841b2f399c5   
6020   540dce6338f0d85bf2515f39   
6417   540a4d6438f0d84fb8833765   
7087   5050cb1a1c22dfde670c6125   
7300   50cba57e00315214fbb7f715   
8454   515dae34cf28d0219d002d7d   
10021  51318342cf28d009de002ffd   
11227  54a1ff2038f0d85064bfc69a   
13180  5565707338f0d84e3574ebfe   
15598  575b5cf238f0d82d6b874796   
17265  589c268e7c459f2525d1b405   

                                                                                                  abstract  \
1045                                                                                                   NaN   
1091                                                                                                   NaN   
1118                                                                                                   NaN   
1219                                                                                                   NaN   
1396                                                                                                   NaN   
3057                                                                                                   NaN   
3575                                                                                                   NaN   
3598                                                                                                   NaN   
7164   Editorial criticizes the Republicans controlling the New York State Senate for not giving a numb...   
45                                                                                                     NaN   
466                                                                                                    NaN   
992                                                                                                    NaN   
995                                                                                                    NaN   
1028                                                                                                   NaN   
1076                                                                                                   NaN   
1174                                                                                             

In [23]:
def text_update(row):
    if row['summary']=='NYTimes.com no longer supports \
            Internet Explorer 9 or earlier.\nPlease upgrade your browser.':
        row['summary']=row['snippet']
    return row

In [24]:
df_clean = df_clean.apply(text_update,axis=1)

In [25]:
df_clean = df_clean[['_id','abstract','headline_clean','keywords','new_desk'\
                     ,'print_page','pub_date','clean_pubdate','year','month','section_name','snippet'\
                    ,'source','type_of_material','web_url','word_count','full_text','summary','author']]

In [26]:
pd.to_pickle(df_clean,"nyt_final")

In [27]:
df_clean.shape

(864, 19)

So we'll have 864 articles to work with